### PR TITLE
chore(tests): migrate to Jest

### PR DESCRIPTION
**Pull request by aaronmcadam**
storybooks-jest

The `storyshots` test is failing. Trying to update the snapshot shows the following error:
storybook-jest-storyshots

### Notes

* I've removed `mocha` TypeScript definitions from `addon-knobs`. Should we replace that with an equivalent for Jest? I'm not a TypeScript user, though I have used flow, so I'm not sure what the answer is here. 
* The `yarn.lock` files still reference `mocha`. [Yarn support is a bit flaky in Lerna](https://github.com/lerna/lerna/issues?utf8=%E2%9C%93&q=yarn) at the moment. I've recently had to stop using it on my Lerna project because I couldn't release. I recommend removing the lock files for now.

### TODO

- [ ] Replace use of `expect` and `chai` with Jest and Enzyme
- [ ] Move Jest configuration to an `rc` file
- [ ] Decide how to test `storyshots`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,16 +1,44 @@
...
   },
   "scripts": {
     "postinstall": "lerna bootstrap",
-    "test": "lerna run test",
-    "lint": "eslint ."
+    "lint": "eslint .",
+    "test": "jest --verbose",
+    "test:watch": "npm test -- --watch"
+  },
+  "jest": {
```
Comment (ndelangen) [inline package.json]: Please move this into a separate config file, just like how babel etc are configured.

File: .eslintrc [hunk 2]
```diff
@@ -1,8 +1,10 @@
 {
   "extends": [
-    "./node_modules/eslint-config-airbnb-base/rules/es6.js"
+    "./node_modules/eslint-config-airbnb-base/rules/es6.js",
+    "plugin:jest/recommended"
```
Comment (ndelangen) [inline .eslintrc]: What does this do?

File: package.json [hunk 3]
```diff
@@ -1,16 +1,44 @@
 {
+  "dependencies": {
+    "react": "^15.4.2",
```
Comment (ndelangen) [inline package.json]: I don't know: should this be peerDependencies && devDependencies?
Or does that only matter in the packages themselves? I'm sure it should be peerDependencies in the packages!

File: package.json [hunk 4]
```diff
@@ -1,16 +1,44 @@
...
+  },
+  "jest": {
+    "cacheDirectory": ".jest",
+    "moduleNameMapper": {
+      "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
+      "\\.(css|scss)$": "<rootDir>/__mocks__/styleMock.js"
+    },
+    "setupTestFrameworkScriptFile": "./node_modules/jest-enzyme/lib/index.js",
+    "snapshotSerializers": [
+      "<rootDir>/node_modules/enzyme-to-json/serializer"
```
Comment (ndelangen) [inline package.json]: I'm assuming this /node_modules/ stuff is unnecessary?

File: package.json [hunk 5]
```diff
@@ -1,16 +1,44 @@
...
+    ],
+    "testResultsProcessor": "./node_modules/jest-junit"
```
Comment (ndelangen) [inline package.json]: What's this for?

File: .eslintrc (hunk 2 above)
Comment (aaronmcadam) [inline .eslintrc]: It sets the recommend rules that Jest provide with their plugin

File: package.json (hunk 3 above)
Comment (aaronmcadam) [inline package.json]: I just leave it defined as a dependency in my root. This dependency is only for the tests at the moment, so I could move it into `dev`

File: package.json (hunk 5 above)
Comment (aaronmcadam) [inline package.json]: I use CircleCI, which needs JUnit XML output. I'm not sure what Travis needs. If it doesn't need anything special for test reporting, I'll remove this

**Comment by aaronmcadam** [comment]
Thanks for the review @ndelangen 

What are your thoughts on my notes above in the PR description?

File: package.json (hunk 4 above)
Comment (aaronmcadam) [inline package.json]: This is needed because we install `enzyme-to-json` in the root.

File: package.json (hunk 1 above)
Comment (aaronmcadam) [inline package.json]: Yeah, I'll sort that at the end

File: package.json (hunk 5 above)
Comment (ndelangen) [inline package.json]: I use jest coverage reporting built-in. This outputs lcov, which I pipe to codecov. This is what I'm planning to do for this repo as well.

File: package.json (hunk 4 above)
Comment (ndelangen) [inline package.json]: Won't this work too:
```
"snapshotSerializers": [
  "enzyme-to-json/serializer"
```

I'm hoping the NodeJS resolver can be used over explicitly stating where you think the file should be.

File: .eslintrc (hunk 2 above)
Comment (ndelangen) [inline .eslintrc]: AH this adds special es-lint rules for jest tests! I ❤️ it!

**Comment by ndelangen** [comment]
> I've removed mocha TypeScript definitions from addon-knobs. Should we replace that with an equivalent for Jest? I'm not a TypeScript user, though I have used flow, so I'm not sure what the answer is here.

I'm no typescript user either, so let's not add it until requested

**Comment by ndelangen** [comment]
> The yarn.lock files still reference mocha. Yarn support is a bit flaky in Lerna at the moment. I've recently had to stop using it on my Lerna project because I couldn't release. I recommend removing the lock files for now.

I'm OK with removing the files adding them to the .gitignore

**Comment by aaronmcadam** [comment]
Closing in favour of https://github.com/storybooks/mono/pull/17

**Pull request closed without merging**